### PR TITLE
Update to Cosmos SDK v0.39.x

### LINKED ISSUE #110: General: Update Cosmos SDK Version to v0.39.x

### PR DESCRIPTION
**Pull request by migueldingli1997**
Closes #110

### CODE REVIEW

File: x/did/auth.go [hunk 1]
```diff
@@ -20,10 +33,10 @@ func GetPubKeyGetter(keeper Keeper) ixo.PubKeyGetter {
 			// For the remaining messages, the did is the signer
 			didDoc, _ := keeper.GetDidDoc(ctx, msg.GetSignerDid())
 			if didDoc == nil {
-				return pubKey, sdk.ErrUnauthorized("Issuer did not found").Result()
+				return pubKey, sdkerrors.Wrap(ErrInvalidDid, "issuer did not found")
```
Comment (stefaniatadama) [inline x/did/auth.go:36]: We should write `DID` instead of `did`

File: x/did/handler.go [hunk 2]
```diff
@@ -3,32 +3,34 @@ package did
...
 		ctx = ctx.WithEventManager(sdk.NewEventManager())
 		switch msg := msg.(type) {
 		case types.MsgAddDid:
 			return handleMsgAddDidDoc(ctx, k, msg)
 		case types.MsgAddCredential:
 			return handleMsgAddCredential(ctx, k, msg)
 		default:
-			return sdk.ErrUnknownRequest("No match for message type.").Result()
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized,
+				"unrecognized did Msg type: %v", msg.Type())
```
Comment (stefaniatadama) [inline x/did/handler.go:23]: We should write `DID` instead of `did`

File: x/did/internal/keeper/querier.go [hunk 3]
```diff
@@ -25,43 +25,43 @@ func NewQuerier(k Keeper) sdk.Querier {
 		case QueryAllDidDocs:
 			return queryAllDidDocs(ctx, k)
 		default:
-			return nil, sdk.ErrUnknownRequest("Unknown did query endpoint")
+			return nil, sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, "Unknown did query endpoint")
```
Comment (stefaniatadama) [inline x/did/internal/keeper/querier.go:28]: We should write `DID` instead of `did`

File: x/did/internal/keeper/querier.go [hunk 4]
```diff
@@ -25,43 +25,43 @@ func NewQuerier(k Keeper) sdk.Querier {
...
+func queryDidDoc(ctx sdk.Context, path []string, k Keeper) ([]byte, error) {
 	didDoc, err := k.GetDidDoc(ctx, path[0])
 	if err != nil {
 		return nil, err
 	}
 
 	res, errRes := codec.MarshalJSONIndent(k.cdc, didDoc)
 	if errRes != nil {
-		return nil, sdk.ErrInternal(fmt.Sprintf("failed to marshal data %s", errRes))
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrJSONUnmarshal, "failed to marshal data %s", errRes.Error())
```
Comment (stefaniatadama) [inline x/did/internal/keeper/querier.go:41]: Should this be `ErrJSONMarshal`?

File: x/did/internal/keeper/querier.go [hunk 5]
```diff
@@ -25,43 +25,43 @@ func NewQuerier(k Keeper) sdk.Querier {
...
 }
 
-func queryAllDids(ctx sdk.Context, k Keeper) ([]byte, sdk.Error) {
+func queryAllDids(ctx sdk.Context, k Keeper) ([]byte, error) {
 	allDids := k.GetAddDids(ctx)
 
 	res, errRes := json.Marshal(allDids)
 	if errRes != nil {
-		return nil, sdk.ErrInternal(fmt.Sprintf("failed to marshal data %s", errRes.Error()))
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrJSONUnmarshal, "failed to marshal data %s", errRes.Error())
```
Comment (stefaniatadama) [inline x/did/internal/keeper/querier.go:52]: Should this be `ErrJSONMarshal`?

File: x/did/internal/types/msgs.go [hunk 6]
```diff
@@ -108,17 +109,17 @@ func (msg MsgAddCredential) String() string {
 		string(msg.DidCredential.Claim.Id), msg.DidCredential.CredType, string(msg.DidCredential.Issuer))
 }
 
-func (msg MsgAddCredential) ValidateBasic() sdk.Error {
+func (msg MsgAddCredential) ValidateBasic() error {
 	// Check if empty
 	if strings.TrimSpace(msg.DidCredential.Claim.Id) == "" {
-		return ErrorInvalidDid(DefaultCodespace, "claim id should not be empty")
+		return sdkerrors.Wrap(ErrInvalidClaimId, "claim ID not be empty")
```
Comment (stefaniatadama) [inline x/did/internal/types/msgs.go:115]: "claim ID **should** not be empty"

Comment (migueldingli1997) [reply x/did/internal/types/msgs.go:115]: Yep, I don't even know why that was changed

File: x/did/internal/keeper/querier.go (hunk 5 above)
Comment (migueldingli1997) [reply x/did/internal/keeper/querier.go:52]: Yes most likely

File: x/did/internal/keeper/querier.go (hunk 4 above)
Comment (migueldingli1997) [reply x/did/internal/keeper/querier.go:41]: Yes most likely

File: x/payments/handler.go [hunk 7]
```diff
@@ -66,35 +67,37 @@ func NewHandler(k Keeper, bk bank.Keeper) sdk.Handler {
 		case MsgEffectPayment:
 			return handleMsgEffectPayment(ctx, k, bk, msg)
 		default:
-			return sdk.ErrUnknownRequest("No match for message type.").Result()
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized,
```
Comment (stefaniatadama) [inline x/payments/handler.go:70]: Maybe `sdkerrors.ErrUnknownRequest` would be better here.

File: x/payments/handler.go [hunk 8]
```diff
@@ -151,39 +154,39 @@ func handleMsgCreatePaymentTemplate(ctx sdk.Context, k Keeper, bk bank.Keeper, m
...
 func handleMsgCreatePaymentContract(ctx sdk.Context, k Keeper, bk bank.Keeper,
-	msg MsgCreatePaymentContract) sdk.Result {
+	msg MsgCreatePaymentContract) (*sdk.Result, error) {
 
 	// Ensure that payment contract doesn't already exist
 	if k.PaymentContractExists(ctx, msg.PaymentContractId) {
-		return types.ErrAlreadyExists(types.DefaultCodespace, fmt.Sprintf(
-			"payment contract '%s' already exists", msg.PaymentContractId)).Result()
+		return nil, sdkerrors.Wrap(types.ErrAlreadyExists, fmt.Sprintf(
+			"payment contract '%s' already exists", msg.PaymentContractId))
```
Comment (stefaniatadama) [inline x/payments/handler.go:166]: We could avoid using `fmt.Sprintf()` here by using `sdkerrors.Wrapf()`, like what is being done in other cases in this file.

File: x/payments/handler.go [hunk 9]
```diff
@@ -151,39 +154,39 @@ func handleMsgCreatePaymentTemplate(ctx sdk.Context, k Keeper, bk bank.Keeper, m
...
-		return sdk.ErrUnauthorized(fmt.Sprintf("%s is not allowed "+
-			"to receive transactions", msg.Payer)).Result()
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "%s is not allowed "+
+			"to receive transactions", msg.Payer)
 	}
 
 	// Confirm that payment template exists
 	if !k.PaymentTemplateExists(ctx, msg.PaymentTemplateId) {
-		return sdk.ErrInternal("invalid payment template").Result()
+		return nil, fmt.Errorf("invalid payment template")
```
Comment (stefaniatadama) [inline x/payments/handler.go:183]: Again, we can probably avoid using `fmt`?

File: x/payments/handler.go [hunk 10]
```diff
@@ -379,38 +383,38 @@ func handleMsgRevokeDiscount(ctx sdk.Context, k Keeper, msg MsgRevokeDiscount) s
...
 	effected, err := k.EffectPayment(ctx, bk, msg.PaymentContractId)
 	if err != nil {
-		return err.Result()
+		return nil, err
 	}
 
 	// Payment not effected but no error, meaning that payment should have been effected
 	if !effected {
-		return sdk.ErrInternal("payment not effected due to unknown reason").Result()
+		return nil, fmt.Errorf("payment not effected due to unknown reason")
```
Comment (stefaniatadama) [inline x/payments/handler.go:417]: Again, we can probably avoid using `fmt`?

File: x/payments/handler.go [hunk 11]
```diff
@@ -285,42 +288,42 @@ func handleMsgCreateSubscription(ctx sdk.Context, k Keeper,
...
 
 	// Confirm that discount ID is in the template (to avoid invalid discount IDs)
 	found, err := k.DiscountIdExists(ctx, contract.PaymentTemplateId, msg.DiscountId)
 	if err != nil {
-		return err.Result()
+		return nil, err
 	} else if !found {
-		return types.ErrInvalidId(types.DefaultCodespace,
-			"discount ID not in payment template's discount list").Result()
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, "discount ID not in payment template's discount list")
```
Comment (stefaniatadama) [inline x/payments/handler.go:320]: Why not do `sdkerrors.Wrap(types.ErrInvalidId,...)`?

File: x/payments/internal/types/subscriptions.go [hunk 12]
```diff
@@ -127,13 +128,13 @@ func (p BlockPeriod) GetPeriodUnit() string {
 	return BlockPeriodUnit
 }
 
-func (p BlockPeriod) Validate() sdk.Error {
+func (p BlockPeriod) Validate() error {
 
 	// Validate period-related values
 	if p.PeriodStartBlock > p.periodEndBlock() {
-		return ErrInvalidPeriod(DefaultCodespace, "start time is after end time")
+		return sdkerrors.Wrap(ErrInvalidId, "start time is after end time")
```
Comment (stefaniatadama) [inline x/payments/internal/types/subscriptions.go:135]: This should be `ErrInvalidPeriod`.

File: x/project/handler.go [hunk 13]
```diff
@@ -30,23 +31,24 @@ func NewHandler(k Keeper, pk payments.Keeper, bk bank.Keeper) sdk.Handler {
...
+			return handleMsgUpdateAgent(ctx, k, msg)
 		case MsgCreateClaim:
 			return handleMsgCreateClaim(ctx, k, msg)
 		case MsgCreateEvaluation:
 			return handleMsgCreateEvaluation(ctx, k, pk, bk, msg)
 		case MsgWithdrawFunds:
 			return handleMsgWithdrawFunds(ctx, k, bk, msg)
 		default:
-			return sdk.ErrUnknownRequest("No match for message type.").Result()
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized,
```
Comment (stefaniatadama) [inline x/project/handler.go:42]: Maybe we could have used `ErrUnknownRequest`?

File: x/project/handler.go [hunk 14]
```diff
@@ -93,45 +94,47 @@ func handleMsgCreateProject(ctx sdk.Context, k Keeper, msg MsgCreateProject) sdk
...
 }
 
 func handleMsgUpdateProjectStatus(ctx sdk.Context, k Keeper, bk bank.Keeper,
-	msg MsgUpdateProjectStatus) sdk.Result {
+	msg MsgUpdateProjectStatus) (*sdk.Result, error) {
 
 	projectDoc, err := k.GetProjectDoc(ctx, msg.ProjectDid)
 	if err != nil {
-		return sdk.ErrUnknownRequest("Could not find Project").Result()
+		return nil, sdkerrors.Wrap(did.ErrInvalidDid, "could not find project")
```
Comment (stefaniatadama) [inline x/project/handler.go:105]: I think we should have used `sdkerrors.ErrUnknownRequest`.

File: x/project/handler.go [hunk 15]
```diff
@@ -153,51 +156,50 @@ func handleMsgUpdateProjectStatus(ctx sdk.Context, k Keeper, bk bank.Keeper,
...
 func payAllFeesToAddress(ctx sdk.Context, k Keeper, bk bank.Keeper, projectDid did.Did,
-	sendingAddress InternalAccountID, receivingAddress InternalAccountID) (sdk.Events, sdk.Error) {
+	sendingAddress InternalAccountID, receivingAddress InternalAccountID) (*sdk.Result, error) {
 	feesToPay := getIxoAmount(ctx, k, bk, projectDid, sendingAddress)
 
 	if feesToPay.Amount.LT(sdk.ZeroInt()) {
-		return nil, sdk.ErrInternal("Negative fee to pay")
-	}
-	if feesToPay.Amount.IsZero() {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidCoins, "negative fee to pay")
```
Comment (stefaniatadama) [inline x/project/handler.go:190]: Not sure `ErrInvalidCoins` is the correct error here.

File: x/project/handler.go [hunk 16]
```diff
@@ -210,18 +212,18 @@ func getIxoAmount(ctx sdk.Context, k Keeper, bk bank.Keeper, projectDid did.Did,
...
 }
 
-func handleMsgCreateAgent(ctx sdk.Context, k Keeper, msg MsgCreateAgent) sdk.Result {
+func handleMsgCreateAgent(ctx sdk.Context, k Keeper, msg MsgCreateAgent) (*sdk.Result, error) {
 
 	// Check if project exists
 	_, err := k.GetProjectDoc(ctx, msg.ProjectDid)
 	if err != nil {
-		return sdk.ErrUnknownRequest("Could not find Project").Result()
+		return nil, sdkerrors.Wrap(did.ErrInvalidDid, "could not find project")
```
Comment (stefaniatadama) [inline x/project/handler.go:220]: We should have used `ErrUnknownRequest`.

File: x/project/auth.go [hunk 17]
```diff
@@ -25,57 +39,53 @@ func GetPubKeyGetter(keeper Keeper, didKeeper did.Keeper) ixo.PubKeyGetter {
...
-				return pubKey, sdk.ErrUnauthorized("signer did not found").Result()
+				return pubKey, sdkerrors.Wrap(sdkerrors.ErrUnauthorized, "signer did not found")
 			}
 			copy(pubKeyEd25519[:], base58.Decode(signerDoc.GetPubKey()))
 		default:
 			// For the remaining messages, the project is the signer
 			projectDoc, err := keeper.GetProjectDoc(ctx, msg.GetSignerDid())
 			if err != nil {
-				return pubKey, sdk.ErrInternal("project did not found").Result()
+				return pubKey, sdkerrors.Wrap(sdkerrors.ErrUnauthorized, "project did not found")
```
Comment (stefaniatadama) [inline x/project/auth.go:49]: Maybe we should use `ErrInvalidDid` instead?

File: x/project/handler.go [hunk 18]
```diff
@@ -402,20 +404,21 @@ func handleMsgCreateEvaluation(ctx sdk.Context, k Keeper, pk payments.Keeper,
...
 	projectDoc, err := k.GetProjectDoc(ctx, withdrawFundsDoc.ProjectDid)
 	if err != nil {
-		return sdk.ErrUnknownRequest("Could not find Project").Result()
+		return nil, sdkerrors.Wrap(did.ErrInvalidDid, "could not find project")
 	}
 
 	if projectDoc.Status != PaidoutStatus {
-		return sdk.ErrUnknownRequest("Project not in PAIDOUT Status").Result()
+		return nil, sdkerrors.Wrap(sdkerrors.ErrUnknownRequest,
+			"project not in PAIDOUT status")
```
Comment (stefaniatadama) [inline x/project/handler.go:421]: In the rest of the file if we get an unexpected project status, the error type is `sdkerrors.ErrUnauthorized`.

File: x/project/internal/keeper/keeper.go [hunk 19]
```diff
@@ -214,13 +217,13 @@ func (k Keeper) ClaimExists(ctx sdk.Context, projectDid did.Did, claimId string)
...
 
-func (k Keeper) GetClaim(ctx sdk.Context, projectDid did.Did, claimId string) (types.Claim, sdk.Error) {
+func (k Keeper) GetClaim(ctx sdk.Context, projectDid did.Did, claimId string) (types.Claim, error) {
 	store := ctx.KVStore(k.storeKey)
 	key := types.GetClaimKey(projectDid, claimId)
 
 	bz := store.Get(key)
 	if bz == nil {
-		return types.Claim{}, did.ErrorInvalidDid(types.DefaultCodespace, "claim not found")
+		return types.Claim{}, fmt.Errorf("claim not found")
```
Comment (stefaniatadama) [inline x/project/internal/keeper/keeper.go:226]: Why not use `sdkerrors.Wrap(did.ErrInvalidDid, "claim not found")`?

File: x/bonds/internal/types/bonds.go [hunk 20]
```diff
@@ -71,17 +72,17 @@ func (fps FunctionParams) Validate(functionType string) sdk.Error {
...
 	// Check that params match and all values are non-negative
 	paramsMap := fps.AsMap()
 	for _, p := range expectedParams {
 		val, ok := paramsMap[p]
 		if !ok {
-			return ErrFunctionParameterMissingOrNonFloat(DefaultCodespace, p)
+			return sdkerrors.Wrapf(ErrFunctionParameterMissingOrNonFloat, p)
 		} else if val.IsNegative() {
-			return ErrArgumentCannotBeNegative(DefaultCodespace, "FunctionParams:"+p)
+			return sdkerrors.Wrapf(ErrArgumentCannotBeNegative, p)
```
Comment (stefaniatadama) [inline x/bonds/internal/types/bonds.go:85]: Maybe `return sdkerrors.Wrapf(ErrArgumentCannotBeNegative, "Function params: %s", p)`?

File: x/bonds/internal/types/errors.go [hunk 21]
```diff
@@ -1,263 +1,53 @@
...
+	ErrDuplicateReserveToken                = sdkerrors.Register(DefaultCodespace, 19, "cannot have duplicate tokens in reserve tokens")
+	ErrInvalidStateForAction                = sdkerrors.Register(DefaultCodespace, 20, "cannot perform that action at the current state")
+	ErrUnrecognizedFunctionType             = sdkerrors.Register(DefaultCodespace, 21, "unrecognized function type")
+	ErrInvalidFunctionParameter             = sdkerrors.Register(DefaultCodespace, 22, "invalid function parameter")
+	ErrFunctionNotAvailableForFunctionType  = sdkerrors.Register(DefaultCodespace, 23, "function is not available for the function type")
+	ErrCannotMakeZeroOutcomePayment         = sdkerrors.Register(DefaultCodespace, 24, "cannot make outcome payment because outcome payment is set to nil")
+	ErrFunctionRequiresNonZeroCurrentSupply = sdkerrors.Register(DefaultCodespace, 25, "function requires the current supply to be non zero")
+	ErrTokenIsNotAValidReserveToken         = sdkerrors.Register(DefaultCodespace, 26, "token is not a valid reserve token")
+	ErrMaxSupplyDenomDoesNotMatchTokenDenom = sdkerrors.Register(DefaultCodespace, 27, "token is not a valid reserve token")
```
Comment (stefaniatadama) [inline x/bonds/internal/types/errors.go:37]: String should be `"max supply denom does not match token denom"`.

File: x/bonds/internal/types/errors.go [hunk 22]
```diff
@@ -1,263 +1,53 @@
...
+	ErrBondTokenCannotAlsoBeReserveToken    = sdkerrors.Register(DefaultCodespace, 28, "token cannot also be a reserve token")
+	ErrBondTokenCannotBeStakingToken        = sdkerrors.Register(DefaultCodespace, 29, "bond token cannot be staking token")
+	ErrReserveDenomsMismatch                = sdkerrors.Register(DefaultCodespace, 30, "denoms do not match reserve denoms")
+	ErrCannotMintMoreThanMaxSupply          = sdkerrors.Register(DefaultCodespace, 31, "cannot mint more tokens than the max supply")
```
Comment (stefaniatadama) [inline x/bonds/internal/types/errors.go:41]: Forgot `ErrInvalidCoinDenomination`?

File: x/bonds/internal/types/msgs.go [hunk 23]
```diff
@@ -73,31 +74,31 @@ func NewMsgCreateBond(token, name, description string, creatorDid did.Did,
...
-		return ErrArgumentCannotBeEmpty(DefaultCodespace, "Function type")
+		return sdkerrors.Wrap(ErrArgumentCannotBeEmpty, "function type")
 	}
 	// Note: FunctionParameters can be empty
 
 	// Check that bond token is a valid token name
 	err := CheckCoinDenom(msg.Token)
 	if err != nil {
-		return ErrInvalidCoinDenomination(DefaultCodespace, msg.Token)
+		return err
```
Comment (stefaniatadama) [inline x/bonds/internal/types/msgs.go:101]: `ErrInvalidCoinDenomination`?

File: x/bonds/internal/types/msgs.go [hunk 24]
```diff
@@ -199,22 +200,22 @@ func NewMsgEditBond(token, name, description, orderQuantityLimits, sanityRate,
...
 
-func (msg MsgEditBond) ValidateBasic() sdk.Error {
+func (msg MsgEditBond) ValidateBasic() error {
 	// Check if empty
 	if strings.TrimSpace(msg.BondDid) == "" {
-		return ErrArgumentCannotBeEmpty(DefaultCodespace, "BondDid")
+		return sdkerrors.Wrap(ErrArgumentCannotBeEmpty, "bond DID")
 	} else if strings.TrimSpace(msg.Token) == "" {
-		return ErrArgumentCannotBeEmpty(DefaultCodespace, "Token")
+		return sdkerrors.Wrap(ErrArgumentCannotBeEmpty, "name")
```
Comment (stefaniatadama) [inline x/bonds/internal/types/msgs.go:208]: `"token"` not `"name"`

File: x/bonds/internal/types/msgs.go [hunk 25]
```diff
@@ -199,22 +200,22 @@ func NewMsgEditBond(token, name, description, orderQuantityLimits, sanityRate,
...
 	} else if strings.TrimSpace(msg.Name) == "" {
-		return ErrArgumentCannotBeEmpty(DefaultCodespace, "Name")
+		return sdkerrors.Wrap(ErrArgumentCannotBeEmpty, "description")
```
Comment (stefaniatadama) [inline x/bonds/internal/types/msgs.go:210]: `"name"` not `"description"`

File: x/bonds/internal/types/msgs.go [hunk 26]
```diff
@@ -199,22 +200,22 @@ func NewMsgEditBond(token, name, description, orderQuantityLimits, sanityRate,
...
 	} else if strings.TrimSpace(msg.Description) == "" {
-		return ErrArgumentCannotBeEmpty(DefaultCodespace, "Description")
+		return sdkerrors.Wrap(ErrArgumentCannotBeEmpty, "sanity rate")
```
Comment (stefaniatadama) [inline x/bonds/internal/types/msgs.go:212]: `"description"` not `"sanity rate"`

File: x/bonds/internal/types/types.go [hunk 27]
```diff
@@ -30,43 +31,43 @@ func CheckReserveTokenNames(resTokens []string, token string) sdk.Error {
 	return nil
 }
 
-func CheckNoOfReserveTokens(resTokens []string, fnType string) sdk.Error {
+func CheckNoOfReserveTokens(resTokens []string, fnType string) error {
 	// Come up with number of expected reserve tokens
 	expectedNoOfTokens, ok := NoOfReserveTokensForFunctionType[fnType]
 	if !ok {
-		return ErrUnrecognizedFunctionType(DefaultCodespace)
+		return sdkerrors.Wrap(ErrUnrecognizedFunctionType, DefaultCodespace)
```
Comment (stefaniatadama) [inline x/bonds/internal/types/types.go:38]: I think instead of `DefaultCodespace` we should have `fnType`.

File: app/app_test.go [hunk 28]
```diff
@@ -6,33 +6,50 @@ import (
...
+		require.Equal(t, !allowedReceivingModAcc[acc], app.BankKeeper.BlacklistedAddr(app.SupplyKeeper.GetModuleAddress(acc)))
 	}
 }
 
+func TestGetMaccPerms(t *testing.T) {
+	dup := GetMaccPerms()
+	require.Equal(t, maccPerms, dup, "duplicated module account permissions differed from actual module account permissions")
+}
+
 func setGenesis(ixoApp *ixoApp) error {
```
Comment (stefaniatadama) [inline app/app_test.go:53]: Maybe we can remove or change `setGenesis()`? The same functionality is now being done in `TestIxodExport()`.

File: x/project/handler.go (hunk 14 above)
Comment (migueldingli1997) [reply x/project/handler.go:106]: Not sure

File: x/project/handler.go (hunk 15 above)
Comment (migueldingli1997) [reply x/project/handler.go:191]: Not sure

Comment (migueldingli1997) [reply x/project/handler.go:191]: Let's leave this as is

File: x/project/handler.go (hunk 14 above)
Comment (migueldingli1997) [reply x/project/handler.go:106]: Let's leave this as is

File: app/app_test.go (hunk 28 above)
Comment (migueldingli1997) [reply app/app_test.go]: Removed it

File: x/bonds/internal/types/types.go (hunk 27 above)
Comment (migueldingli1997) [reply x/bonds/internal/types/types.go:38]: Great catch

File: x/bonds/internal/types/msgs.go (hunk 26 above)
Comment (migueldingli1997) [reply x/bonds/internal/types/msgs.go:212]: Correct

File: x/bonds/internal/types/msgs.go (hunk 25 above)
Comment (migueldingli1997) [reply x/bonds/internal/types/msgs.go:210]: Correct

File: x/bonds/internal/types/msgs.go (hunk 24 above)
Comment (migueldingli1997) [reply x/bonds/internal/types/msgs.go:208]: Correct

File: x/bonds/internal/types/errors.go (hunk 21 above)
Comment (migueldingli1997) [reply x/bonds/internal/types/errors.go:37]: Correct

File: x/project/internal/keeper/keeper.go (hunk 19 above)
Comment (migueldingli1997) [reply x/project/internal/keeper/keeper.go:226]: Claims are not identified by DIDs and we do not have a custom error type for invalid claims.

File: x/project/handler.go (hunk 18 above)
Comment (migueldingli1997) [reply x/project/handler.go:422]: Correct

File: x/project/auth.go (hunk 17 above)
Comment (migueldingli1997) [reply x/project/auth.go:49]: Correct

File: x/project/internal/keeper/keeper.go (hunk 19 above)
Comment (migueldingli1997) [reply x/project/internal/keeper/keeper.go:226]: Will leave it as is

File: x/bonds/internal/types/msgs.go (hunk 23 above)
Comment (migueldingli1997) [reply x/bonds/internal/types/msgs.go:101]: We can just use the error that CheckCoinDenom is returning

**Review by stefaniatadama — APPROVED** [review]

**Pull request merged**